### ISSUE TITLE
New Password UI/UX

### ISSUE THREAD

**Issue by toby63** [vector-im/element-meta#1947]
#### What would you like to do?

Many of my friends struggle with the different passwords (by which I mean mainly the password for servers and the recovery key etc.), so I would like to propose some different solutions:

First of all regarding the initial generation of keys:

I would like to have the recovery key created/set automatically at the beginning (first login), so it is not some optional, complicated thing that most users (in my experience) don't understand.
Think of it more like a second password, that users have to write down.

Secondly regarding the login:

1. One-Password Solution:
  The idea is quite simple. You technically still have two passwords, but they get combined into one for the users.
So imagine you login a second time (new session/new device); instead of having to type in two or more passwords in different menus, you simply type in one password, which gets seperated and processed by the client behind the scenes.

2. Two Password fields in one window/menu:
  This idea is very similar to the idea about initial generation above; instead of having seperate menus, you simply have one:
```
     Login Username:
     Password1 aka Server password:
     Password2 aka recovery key:
```

### Have you considered any alternatives?

-

### Additional context

-

**Comment by t3chguy** [comment]
> you simply type in one password, which gets seperated and processed by the client behind the scenes.

This isn't possible as Matrix supports auth systems like LDAP, so you need to send the original password to the server.

**Comment by toby63** [comment]
> > you simply type in one password, which gets seperated and processed by the client behind the scenes.
> 
> This isn't possible as Matrix supports auth systems like LDAP, so you need to send the original password to the server.

I don't really understand that.
As I wrote in the exact sentence you quoted, the original password still exists, but the user types in a combined password and the client kind of extracts the original password from it, so I don't see a problem.
Even for different kinds of verification, because the second part of the password could simply be ignored by the client.

**Comment by t3chguy** [comment]
With AD/LDAP the password the user enters matches between different systems. Given that when a user logs into Windows or their email they'll be inputting $password.

Element could take $password and use a hash with 2 different salts to generate $derived1 and $derived2 where the former is used for auth and the latter for crypto. But $derived1 != $password so would fail LDAP.
If Element just used $password as the Matrix password (allowing LDAP) and a cryptographically derived password for verification then the server would also know your verification key due to being able to capture your password during authentication and apply the same hash, thus breaking end to end encryption.

> but the user types in a combined password and the client kind of extracts the original password from it, so I don't see a problem.

This is not how shared-auth systems like LDAP work. The user has one common password between all their [corporate;related] services, changing password in one place changes it in all places.

**Comment by toby63** [comment]
I might not see something here, but my thinking was extremly simple, let me give an example:

In the beginning you set the passwords, e.g.:
```
serverpassword=peterpan523
recoverykey=39594jfjb6238587492    # which is generated diffently I know, but this is only an example
```

Now the idea is that the user will be enabled to type in these passwords combined in the client on login, so:

`user types in: peterpan523_39594jfjb6238587492`

The "_" indicates the seperation between the two passwords.

So the client reads all passwords just like above and can process and send them further.

From my point of view it's just a UI thing, nothing changes in the background. But as I said, maybe I don't understand the technology :smile:.

**Comment by t3chguy** [comment]
That would be very different to everything else a user in an LDAP set up would be used to. They just use the very same password everywhere. Having to glue a tumour onto the end of their password would confuse them.

> The "_" indicates the seperation between the two passwords.

What if a user's password has symbols such as underscore within their password? Without having a fixed length to either left or right hand side of the separator, or a separate length included somewhere finding the split would not be feasible.

**Comment by toby63** [comment]
> That would be very different to everything else a user in an LDAP set up would be used to. They just use the very same password everywhere. Having to glue a tumour onto the end of their password would confuse them.

:thinking: Well I only wanted the one-password solution as an option, so they could use the regular way of login.
Or Option 2 in my request above.

Alternatively if the 1-password scenario is used, they could also simply type in just the regular password and the client detects that only one password is given and requests other passwords/keys seperately (which of couse brings us to the discussion below).

> 
> > The "_" indicates the seperation between the two passwords.
> 
> What if a user's password has symbols such as underscore within their password? Without having a fixed length to either left or right hand side of the separator, or a separate length included somewhere finding the split would not be feasible.

I understand the problem, my thinking was of couse that one specific symbol, in this example the underscore, would not be allowed for matrix passwords, but I guess that does not apply to other setups.

**Comment by t3chguy** [comment]
> would not be allowed for matrix passwords, but I guess that does not apply to other setups.

And immediately renders you incompatible with shared auth systems like LDAP, where the password requirements are wholly external to the Matrix system

**Comment by toby63** [comment]
> > would not be allowed for matrix passwords, but I guess that does not apply to other setups.
> 
> And immediately renders you incompatible with shared auth systems like LDAP, where the password requirements are wholly external to the Matrix system

True, so it could only be optional, or with a checkbox or something indicating the use of LDAP.

Anyway the point of my request is still making the whole password and key thing a bit easier, because most of my friends, colleagues and family members simply ignore that and then they get into problems like:
"I can't recover my past sessions" (because they did not generate a recovery key etc. and they don't get what that is and complain about unnecessary complexety etc.)

**Comment by t3chguy** [comment]
Matrix is moving to OIDC where you enter your password on a website controlled by your server, so such a proposal will not work there

**Comment by toby63** [comment]
> Matrix is moving to OIDC where you enter your password on a website controlled by your server, so such a proposal will not work there

And how is the recovery key kept secret from the server in such a scenario?

**Comment by t3chguy** [comment]
It is entered in the client, hence why you cannot join password & recovery key into one string given there'll be two UIs each controlled by a different entity you enter them into

**Comment by toby63** [comment]
> It is entered in the client, hence why you cannot join password & recovery key into one string given there'll be two UIs each controlled by a different entity you enter them into

Very unfortunate.
And the "Two Password fields in one window/menu" option is also impossible?

**Comment by t3chguy** [comment]
In OIDC you input the password in a webpage of your homeserver, e.g. matrix.org instead of your client e.g. app.element.io. Like when you use "Sign in with Google" you input your google credentials into https://accounts.google.com/ rather than whatever you're trying to use. Adding your recovery field there, even if possible, would break end-to-end encryption by leaking your secrets to your server.

**Comment by toby63** [comment]
Well the idea would not be that of course, but depending on how the webpages look, it would be possible to have iframes on one independent offline page which includes different webpages.

Also I assume that there must be a way for non-web clients to implement it without the webpage or not?

**Comment by t3chguy** [comment]
No. OIDC often blocks iframing due to clickjacking concerns. Given any OIDC provider must be supported iframing cannot be relied upon.

> Also I assume that there must be a way for non-web clients to implement it without the webpage or not?

OIDC has alternate flows for non-web clients, like how you might log into Netflix on your TV with a code you get on the TV and you enter into the service on a web browser-enabled device. But your credentials are only ever entered into the OIDC service, never directly into a client.

**Comment by toby63** [comment]
> No. OIDC often blocks iframing due to clickjacking concerns. Given any OIDC provider must be supported iframing cannot be relied upon.

Understandable.

> 
> > Also I assume that there must be a way for non-web clients to implement it without the webpage or not?
> 
> OIDC has alternate flows for non-web clients, like how you might log into Netflix on your TV with a code you get on the TV and you enter into the service on a web browser-enabled device. But your credentials are only ever entered into the OIDC service, never directly into a client.

That sounds very complicated and annoying. Is this already decided upon and when will this roll out?

Or did I misunderstand something, and this is only a login option and thus this would "only" block my request for compatibility reasons?

**Comment by t3chguy** [comment]
https://matrix.org/blog/2022/08/10/areweoidcyet-com-matrix-and-open-id-connect/

Both login options would be available but for some time but going forward OIDC-only clients would be the case to simplify the code and security harden it.

**Comment by toby63** [comment]
Very interesting, thx for the information.

But all this does not sound like anything is going to be easier (I hope you (the matrix team) have thought about this well, because some users and server admins (including me) will not like this at all, I hope you at least include an option for not having to use 2FA). This might cause some to leave Element and Matrix behind.

So to summarize: aside from some possible changes in wording to make things easier and better understandable, none of my requests will be integrated.

**Comment by t3chguy** [comment]
OIDC has now landed on matrix.org and many other servers so a solution of gluing the passwords or showing both entry fields at once is not possible
